Exctract code from WPML_PB_Reuse_Translations and WPML_PB_String_Translation to a parent class

strategio opened the pull request:
This change is required for the Gutenberg package (https://github.com/OnTheGoSystems/wpml-page-builders-gutenberg/pull/27).

Actually, I did as follow:
- Rename the existing classes with '_By_Strategy'
- Create new classes with the original names
- Move base code to the new classes
- Make '_By_Strategy' classes extend the new classes

https://onthegosystems.myjetbrains.com/youtrack/issue/wpmlcore-6326

BruceOnTheGo commented:
@strategio It's pretty hard to review with Scrutinizer complaining all the way through. I think some of the issues are valid. Can you fix those issues.

Maybe the base classes should also be abstract if they are only to be extended.

strategio commented:
@brucepearson, I will address the complaints of Scrutinizer. But I will also implement the counter part on Gutenberg package so it will be easier to understand (and I was not planning to use the new classes as abstract). Please wait...

strategio commented:
@brucepearson, this is ready for review now. You can check how the classes are used in https://github.com/OnTheGoSystems/wpml-page-builders-gutenberg/pull/27.

strategio commented:
@OnTheGoSystems/wpml please review with https://github.com/OnTheGoSystems/wpml-page-builders-gutenberg/pull/27.

strategio commented:
@OnTheGoSystems/wpml one more review please :)

strategio commented:
@OnTheGoSystems/wpml please one more review!

andreasciamanna commented on src/st/class-wpml-pb-reuse-translations.php:
```diff
@@ -68,10 +63,10 @@ private function find_existing_strings_for_new_strings( array $new_strings, arra
+	 * @param array[] $leftover_strings
 	 *
-	 * @return array
+	 * @return array[]
```
@strategio Scrutinizer here has a point: why you can't fix this inspection issue?

strategio replied:
Here's an example of the array @andreasciamanna:

```
array(
    'gf4544ds454sds542122sd' => array(
		'value' 	 => 'The string value',
		'context'    => 'the-string-context',
		'name'       => 'the-string-name',
		'id'         => 123,
		'package_id' => 123,
		'location'   => 123,
	),
)
```

I am not sure it will help to improve the PHPDoc, but I will add this structure as a comment in the main method.

Also, I already fixed this PHPDoc once after Scrutinizer was complaining, and now it comes with another suggestion.

andreasciamanna replied:
@strategio I thought it was just an array of strings (`string[]`). If it's more complicated, forget it :)

strategio commented:
I removed one WIP commit which was included while I added the last fixup for Scrutinizer. It was an attempt to go one mile further in the refactorization and it does not work.
